enableclangd

kirkshoop opened the pull request:
- **spelling fix**
- **icwyu fixes**
- **atomic type fixes**
- **enable clangd**
- **add let_error and let_stopped**
- **fix open/close cancellation tests**

kirkshoop committed:
spelling fix

kirkshoop committed:
icwyu fixes

kirkshoop committed:
atomic type fixes

kirkshoop committed:
enable clangd

kirkshoop committed:
add let_error and let_stopped

kirkshoop committed:
fix open/close cancellation tests

kirkshoop commented:
It took me a while to get squiz building. I added clangd support that works well with vscode.

The atomic changes were made to remove compile errors 'cant initialize const' for atomic_ref load and mismatch for volatile from the atomic_ref constructor. These might need a different fix.

I was playing around with the open/close tests and they did not seem to actually cancel and once cancelling, would not close the handle. I ended up adding let_error and let_stopped to get the behavior that I expected.

lewissbaker commented on .clangd:
```diff
@@ -0,0 +1,30 @@
+# https://clangd.llvm.org/config
+
+# Apply a config conditionally to all C++ files
```
It doesn't seem like this part of the config is doing anything?
You are matching C++ files and then not applying any changes to the build config for them. Is that right?
All of the settings are set unconditionally for all files in the next section.

kirkshoop replied:
Yep. removed.

lewissbaker commented on include/squiz/linuxos/io_uring.hpp:
```diff
@@ -410,19 +408,19 @@ class io_uring {
   io_uring_sqe* sq_entries_;
   std::uint32_t* sq_index_array_;
-  volatile const std::uint32_t* sq_head_;
+  std::uint32_t* sq_head_;
```
Why was the 'volatile' qualifier removed here?

The intent of the use of 'volatile' was to indicate to the compiler that this memory
may change spuriously for reasons outside of this program (e.g. because the kernel
wrote to this memory). I have tried to tag all memory that is mapped in via shared
memory with the 'volatile' qualifier.

lewissbaker commented on source/linux/io_uring.cpp:
```diff
@@ -336,18 +336,18 @@ bool io_uring::supports_completion_queue_nodrop() const noexcept {
   std::uint32_t tail = sqe_tail_;
-  std::atomic_ref<const volatile std::uint32_t> head_ref{*sq_head_};
-  std::uint32_t head = head_ref.load(std::memory_order_acquire);
+  std::atomic_ref<std::uint32_t> head_ref{*sq_head_};
```
Ah, I think I had made some local hack changes to my libc++ to support `atomic_ref<volatile T>`, which may be why I didn't have issues with the volatile-qualified `sq_head_` pointer here.

lewissbaker replied:
Hopefully libc++ will support https://wg21.link/p3323 soon.

kirkshoop replied:
I added a comment to reintroduce volatile once P3323 is implemented.

lewissbaker commented on test/source/linux/iouring_context.cpp:
```diff
@@ -264,16 +264,23 @@ TEST_CASE("iouring_context - openat/close") {
-                }};
+                  std::puts("opened /dev/zero");
+                  return squiz::let_value_sender{
+                      sched.schedule_after(100ms),
```
Was there an issue with immediately closing the file after it was opened?
Or did you just want to have more of a delay between the open and close actions?

kirkshoop replied:
the immediate close prevented any possibility of cancellation. I went ahead and removed the delay for the basic success test.

lewissbaker commented on cmake/clangd_compile_info.cmake:
```diff
@@ -0,0 +1,30 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+include(${CMAKE_CURRENT_SOURCE_DIR}/../cmake/cmake_command.cmake)
```
`${CMAKE_CURRENT_SOURCE_DIR}/../cmake` is just `${CMAKE_CURRENT_SOURCE_DIR}`, right?

Can this be simplified to just:
```cmake
include(${CMAKE_CURRENT_SOURCE_DIR}/cmake_command.cmake)
```

kirkshoop replied:
No, `${CMAKE_CURRENT_SOURCE_DIR}` is where the CMakeLists.txt is - This is troublesome in this project since the `all` and `test` subdirs are each 'top-level' projects. Thus  `cmake/clangd_compile_info.cmake` can only be included from `all` and `test`. Only `all` includes it so only `all` builds get clangd. I did not find any way to make it work regardless of which project included it.

lewissbaker commented on include/squiz/let_value.hpp:
```diff
@@ -25,51 +25,51 @@
-template <typename Source, typename BodyFactory, typename Receiver>
-using let_value_child_op = value_signatures_t<
+template <typename ResultTag, typename Source, typename BodyFactory, typename Receiver>
+using let_common_child_op = value_signatures_t<
```
This doesn't quite look right.

Shouldn't this be extracting the signatures for the specified `ResultTag` here rather than the `value_signatures_t`?

e.g.
```c++
template <typename ResultTag, typename Source, typename BodyFactory, typename Receiver>
using let_common_child_op = transform_signatures_t<
  completion_signatures_for_t<Source, receiver_env_t<Receiver>>,
  only_tags<ResultTag>>::template apply<
    make_let_common_variant_child_operation<ResultTag, Source, BodyFactory, Receiver>::
      template apply>;
```

lewissbaker commented on include/squiz/let_value.hpp:
```diff
@@ -25,51 +25,51 @@
-using let_value_transform_t =
-    typename let_value_transform<Sig, BodyFactory>::type;
+using let_common_transform_t =
+    typename let_common_transform<Sig, BodyFactory>::type;
```
I  think this now also needs to accept the 'Tag' type so that it can extract and apply the body factory to the corresponding signature.
e.g.
```
template<typename ResultTag, typename Sig, typename BodyFactory>
struct let_common_transform {};

template<typename ResultTag, typename... Datums, typename BodyFactory>
  requires std::invocable<BodyFactory, Datums&...>
struct let_common_transform<result_t<ResultTag, Datums...>, BodyFactory> {
  using type = std::invoke_result_t<BodyFactory, Datums&...>;
};

template<typename ResultTag, typename Sig, typename BodyFactory>
using let_common_transform_t =
  typename let_common_transform<ResultTag, Sig, BodyFactory>::type;
```
and then update usage below in `make_let_common_variant_child_operation` to use `let_common_transform_t<ResultTag, Sigs, BodyFactory>`.

kirkshoop replied:
Got it. I think this is fixed now.

lewissbaker commented on include/squiz/let_value.hpp:
```diff
@@ -108,33 +108,33 @@ struct transform_source_signatures {
   template <typename... Vs>
     requires std::invocable<BodyFactory, Vs&...>
-  static auto apply(value_t<Vs...>) -> body_completion_sigs_t<Vs...>;
+  static auto apply(result_t<ResultTag, Vs...>) -> body_completion_sigs_t<Vs...>;
```
When dealing with a generic `result_t` rather than a `values_t<...>` result I generally try to use the term `Datums` instead of `Vs` or `Values`.

Can we rename the template params from `Vs` to `Datums` here (and in the helpers above)?

lewissbaker commented on include/squiz/let_value.hpp:
```diff
@@ -221,7 +221,7 @@ class let_value_op final
   template <typename... Vs>
```
Similarly, here. Can we rename the template param from `Vs` to `Datums` since they can now refer to either values or errors?

Similarly for the `vs` function parameter name.

lewissbaker commented on include/squiz/let_value.hpp:
```diff
@@ -237,23 +237,25 @@ class let_value_op final
     }
 
     try {
       auto& value_tuple =
```
Rename this variable to `result_tuple`?

lewissbaker commented on include/squiz/let_value.hpp:
```diff
@@ -237,23 +237,25 @@ class let_value_op final
-      // set_stopped().
-      squiz::set_stopped(this->get_receiver());
-      return;
+    if constexpr (one_of<ResultTag, value_tag, error_tag>) {
```
It's an interesting question of whether or not to run the body-function of `let_stopped` if the source completes with 'stopped' if there is an outstanding stop-request.

Here we decide to run the BodyFunction anyway, but down below on line 286, we decide to propagate the stop-request to immediately send a stop-request to the body-sender.

If we are handling a stop-result, does it make sense to have operations performed by that stop-result also stopped? Or should `let_stopped` be implicitly blocking stop-requests from propagating to the body of a `let_stopped`?

kirkshoop replied:
I intended to run the body so that resource cleanup could be composed and run even in the presence of cancellation.
I added a test case to show this, and discovered that the body needed to be unconditionally run for let_value and let_error as well, so I removed the whole statement.

lewissbaker reviewed (changes requested):
Thanks for the improvements, @kirkshoop !

I added a few comments about a bug and a few minor issues, but otherwise looks good.

kirkshoop committed:
apply feedback: renames, let_.. result calculation, etc..
Added a preemptive cancellation test case to demo the need to run the body factory unconditionally in order to compose unstoppable resource cleanup, like close().

kirkshoop reviewed (commented):
> Hopefully libc++ will support https://wg21.link/p3323 soon. 

I added a comment to reintroduce volatile once P3323 is implemented.
